svd_utils.py

```
@@ -117,35 +270,279 @@ def project_gradient_to_orthogonal_space(svd_dict):
+
+    Args:
+        model: The model with SVD parameters
+        assignments: List of parameter assignments for each rank
+        world_size: Number of distributed processes
+    """
+    # Create a mapping from parameter name to the rank that computed it
+    param_to_rank = {}
+    for rank, params in enumerate(assignments):
+        for name, _ in params:
+            param_to_rank[name] = rank
+
+    # Broadcast each parameter from its computing rank
+    for name in param_to_rank:
+        src_rank = param_to_rank[name]
+        safe_name = name.replace(".", "_")
+
+        # Broadcast buffer components (high-rank frozen components)
+        buffer_components = [
+            f"{safe_name}_U_high",
+            f"{safe_name}_S_high",
+            f"{safe_name}_V_high",
+        ]
+
+        for component_name in buffer_components:
+            try:
+                if hasattr(model, component_name):
+                    tensor = getattr(model, component_name)
+                    dist.broadcast(tensor, src=src_rank)
+                else:
+                    print(f"Warning: Buffer {component_name} not found in model")
+            except Exception as e:
+                print(f"Warning: Failed to broadcast buffer {component_name}: {e}")
+                continue
+
+        # Broadcast trainable components (low-rank parameters)
+        try:
+            if safe_name in model.svd_params:
+                svd_module = model.svd_params[safe_name]
+
+                # Broadcast U_low, S_low, V_low
+                dist.broadcast(svd_module.U_low, src=src_rank)
+                dist.broadcast(svd_module.S_low, src=src_rank)
+                dist.broadcast(svd_module.V_low, src=src_rank)
+            else:
+                print(f"Warning: SVD module {safe_name} not found in model.svd_params")
+        except Exception as e:
+            print(f"Warning: Failed to broadcast SVD module {safe_name}: {e}")
+            continue
```

Should this fail silently? Shouldn't we crash at this point?

---

Good catch, this should fail explicitly.